src/Compiler/Microsoft.AspNetCore.Razor.Language/src/RazorHtmlWriter.cs

```diff
@@ -89,7 +92,7 @@ private RazorHtmlWriter(RazorSourceDocument source)
 
         var sourceMappings = writer.SourceMappings.ToArray();
 
-        var razorHtmlDocument = new DefaultRazorHtmlDocument(codeDocument, generatedHtml, options, sourceMappings);
+        var razorHtmlDocument = new DefaultRazorHtmlDocument(codeDocument, generatedHtml, options, sourceMappings.ToImmutableArray());
```

Comment:
Tiny nit: Since you have a builder in hand, you can call the `ToImmutable()` instance method rather than the `ToImmutableArray()` extension method, which adds an unneeded null check.

Reply:
Thank you for this comment. There was no `ToImmutable` method in completion, and it made me realise we both missed the real crime which is two lines above this where the immutable array is converted to an array 🤦‍♂️

Reply:
Actually.. is DrainToImmutable better? I think thats the good one, right?

Reply:
Sure. When you don't know and haven't set the capacity ahead of time, `DrainToImmutable()` has a small chance of avoiding an allocation if the capacity of the builder is the same as the count.